src/vscode_ocaml_platform.ml

```diff
@@ -1,263 +1,69 @@
-  | None -> ()
-  | Some () -> select_sandbox instance ()
-
-let register_commands extension =
-  List.iter ~f:(function command, callback ->
-      let callback ~args:_ = callback () in
-      ExtensionContext.subscribe extension
-        ~disposable:(Commands.registerCommand ~command ~callback))
+  Option.iter selection ~f:(Extension_commands.select_sandbox.handler instance)
 
 let activate (extension : ExtensionContext.t) =
+  (* this env var update disables ocaml-lsp's logging to a file
+     because we use vscode [output] pane for logs *)
   Process.Env.set "OCAML_LSP_SERVER_LOG" "-";
-  let instance = Instance.create () in
-  register_commands extension
-    [ (select_sandbox_command_id, select_sandbox instance)
-    ; (restart_command_id, restart_instance instance)
-    ; (open_terminal_command_id, open_terminal instance)
-    ; (open_terminal_select_command_id, select_sandbox_and_open_terminal)
-    ; (switch_impl_intf_command_id, switch_impl_intf instance)
-    ];
-  ExtensionContext.subscribe extension
-    ~disposable:(Instance.disposable instance);
   let open Promise.Syntax in
-  let toolchain =
-    let+ pm = Toolchain.of_settings () in
-    let resources, is_fallback =
-      match pm with
-      | Some toolchain -> (toolchain, false)
-      | None ->
-        let (_ : unit Promise.t) = suggest_to_setup_toolchain instance in
-        (Toolchain.Package_manager.Global, true)
-    in
-    (Toolchain.make_resources resources, is_fallback)
+  let* package_manager =
+    Toolchain.of_settings ()
+    (* TODO: implement [Toolchain.from_settings_or_detect] that would
+       either get the sandbox from the settings or detect in a smart way (not simply Global) *)
+  in
+  let is_fallback = Option.is_empty package_manager in
+  let package_manager =
+    Option.value package_manager ~default:Toolchain.Package_manager.Global
   in
-  let* toolchain, is_fallback = toolchain in
-  Instance.start instance toolchain
-  |> Promise.Result.iter ~error:(fun e ->
-         if not is_fallback then message `Error "%s" e)
+  Extension_instance.make (Toolchain.make package_manager)
+  |> Promise.Result.iter
+       ~ok:(fun instance ->
+         (* register things with vscode, making sure to register their disposables *)
+         let _register_extension_commands : unit =
+           Extension_commands.register_all_commands extension instance
+         in
+         let _register_extension_instance : unit =
+           ExtensionContext.subscribe extension
+             ~disposable:(Extension_instance.disposable instance)
+         in
+         let _register_dune_formatter : unit =
+           Dune_formatter.register instance
+           |> List.iter ~f:(fun disposable ->
+                  ExtensionContext.subscribe extension ~disposable)
+         in
+         let _register_dune_task_provider : unit =
+           let disposable = Dune_task_provider.register instance in
+           ExtensionContext.subscribe extension ~disposable
+         in
```

Comment:
I don't see the point of giving names to these bindings if they're going to be unused. I would just separate the statements with semicolons.

Reply:
you're right, but with names it looks a little more structured and declarative; if not bindings, I would at least comment these, but again with bindings it looks more structured

```ocaml
         let _register_extension_commands : unit =
           Extension_commands.register_all_commands extension instance
         in
         let _register_extension_instance : unit =
           ExtensionContext.subscribe extension
             ~disposable:(Extension_instance.disposable instance)
         in
         let _register_dune_formatter : unit =
           Dune_formatter.register instance
           |> List.iter ~f:(fun disposable ->
                  ExtensionContext.subscribe extension ~disposable)
         in
         let _register_dune_task_provider : unit =
           let disposable = Dune_task_provider.register instance in
           ExtensionContext.subscribe extension ~disposable
```

vs 

```ocaml
           Extension_commands.register_all_commands extension instance;
           ExtensionContext.subscribe extension
             ~disposable:(Extension_instance.disposable instance);
           Dune_formatter.register instance
           |> List.iter ~f:(fun disposable ->
                  ExtensionContext.subscribe extension ~disposable);
           let disposable = Dune_task_provider.register instance in
           ExtensionContext.subscribe extension ~disposable;
```

Reply:
With some API changes, you can write
```ocaml
Extension_commands.register_all_commands extension instance;

ExtensionContext.subscribe extension
    ~disposable:(Extension_instance.disposable instance);
    
let dune_formatter = Dune_formatter.register instance in
ExtensionContext.subscribe extension ~disposable:dune_formatter in

let dune_task_provider = Dune_task_provider.register instance in
ExtensionContext.subscribe extension ~disposable:dune_task_provider;
```

Which is perfectly readable in my opinion.